### PR TITLE
Hierarchical components

### PR DESCRIPTION
**Pull request by pmlodawski**
This pull request introduces new approach to create gui components. For now it is implemented only for connections, but it is backward compatible, co we can move more components into that. This also fixes problems solving in #5

### CODE REVIEW

File: src/view/BasicComponent.coffee [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    __undraw: (def) => @withScene (scene) =>
+        if def?
+            scene.delete def
+            @__element = null
+
+    redefineRequred: => false
+
+    onModelUpdate: =>
+        if @redefineRequred(@changed) or (not @__def?)
```
Comment (ljkania) [inline src/view/BasicComponent.coffee]: Change elements order - the second one happens much quicker.

File: src/view/BasicComponent.coffee [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            def = @define()
+            if def? and not _.isEqual(def, @__def)
+                @__undraw @__def
+                @__draw def
+                @registerEvents? @__element
```
Comment (ljkania) [inline src/view/BasicComponent.coffee]: This doesn't use `@__def` but may. Why not assign it first?

File: src/view/Component.coffee [hunk 3]
```diff
@@ -6,6 +6,7 @@ export class Component extends EventEmitter
     cons: (values, @parent) =>
         super()
         @set values
+        @attach()
```
Comment (ljkania) [inline src/view/Component.coffee]: Are you sure we want to do it explicitly here? What is the motivation?

File: src/view/Connection.coffee [hunk 4]
```diff
@@ -18,35 +19,63 @@ connectionShape.variables.color_r = 1
...
 
-export class Connection extends Component
-    constructor: (args...) ->
-        super args...
-        @srcNodeSubscirbed = false
-        @dstNodeSubscribed = false
-        @disposables = []
+class PartialConnection extends BasicComponent
+    initModel: =>
+        offset: 0
```
Comment (ljkania) [inline src/view/Connection.coffee]: This should be not needed here. As we discussed the only thing it needs is `length`, `color` and `src`, rotation and position should be set by parent using `thisElement.view...`

File: src/view/Connection.coffee [hunk 5]
```diff
@@ -18,35 +19,63 @@ connectionShape.variables.color_r = 1
...
+        length: 0
+        angle: 0
```
Comment (ljkania) [inline src/view/Connection.coffee]: ditto

File: src/view/ContainerComponent.coffee [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    def: (key) => @__defs[key]
+
+    addDef: (key, def) =>
+        unless @__defs[key]?
+            @__defs[key] = def
+            @__defsModified = true
+        else unless _.isEqual @__defs[key], def
+            @__defs[key].dispose()
+            @__defs[key] = def
+            @__defsModified = true
```
Comment (ljkania) [inline src/view/ContainerComponent.coffee]:
```coffee
unless _.isEqual @__defs?[key], def
  @__defs?[key]?.dispose()
  @__defs[key] = def
  @__defsModified = true
```
This should work as well.

File: src/view/HasModel.coffee [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        @changed = {}
+        @prepare?()
+        @withScene =>
+            @set values
+            @connectSources?()
+
+    withScene: (fun) => @parent.withScene fun if @parent?
+
+    set: (values) =>
+        return unless values?
```
Comment (ljkania) [inline src/view/HasModel.coffee]: What this line do? #coffeenoob

File: src/view/Component.coffee (hunk 3 above)
Comment (pmlodawski) [reply src/view/Component.coffee]: This is only to make old Components compatible

Comment (ljkania) [reply src/view/Component.coffee]: Please comment.

File: src/view/BasicComponent.coffee (hunk 1 above)
Comment (pmlodawski) [reply src/view/BasicComponent.coffee]: Not agree. In most cases `@__def` will be null only once, at initialisation

File: src/view/HasModel.coffee (hunk 7 above)
Comment (pmlodawski) [reply src/view/HasModel.coffee]:
```
unless values?
    return
```

File: src/view/BasicComponent.coffee (hunk 1 above)
Comment (ljkania) [reply src/view/BasicComponent.coffee]: Yup, but redefineRequred may be pretty expensive. `not @__def?` is never expensive.

File: src/view/Component.coffee (hunk 3 above)
Comment (pmlodawski) [reply src/view/Component.coffee]: But it simply works, no need to attach them manually ;)

File: src/view/ContainerComponent.coffee (hunk 6 above)
Comment (pmlodawski) [reply src/view/ContainerComponent.coffee]: Yes, but I'm not sure it is more readable...

File: src/view/BasicComponent.coffee (hunk 2 above)
Comment (pmlodawski) [reply src/view/BasicComponent.coffee]: Can you give an example? 
`@__undraw` undraws old `@__def`, 
`@__draw` draws new `def`.

File: src/view/BasicComponent.coffee (hunk 1 above)
Comment (pmlodawski) [reply src/view/BasicComponent.coffee]: But will be called everytime component is changed and hardly never will be null

File: src/view/BasicComponent.coffee (hunk 2 above)
Comment (ljkania) [reply src/view/BasicComponent.coffee]: Okay, you are right. @registerEvents gets an argument. It is ok.

File: src/view/ContainerComponent.coffee (hunk 6 above)
Comment (ljkania) [reply src/view/ContainerComponent.coffee]: For me it is, but your choice. This one simply says: if there was a change do:
1. if there was `@__def` and it has key `key` do `dispose()` on it
2. Same as yours
3. Same as yours
Just a hint, not a problem.

**Pull request merged**